soc/xtensa/intel_adsp/ace/power.c

```diff
@@ -119,37 +105,13 @@ struct lpsram_header {
 	uint8_t rom_bypass_vectors_reserved[0xC00 - 0x14];
 };
 
-static ALWAYS_INLINE void _save_core_context(uint32_t core_id)
-{
-	core_desc[core_id].vecbase = XTENSA_RSR("VECBASE");
-	core_desc[core_id].excsave2 = XTENSA_RSR("EXCSAVE2");
-	core_desc[core_id].excsave3 = XTENSA_RSR("EXCSAVE3");
-	core_desc[core_id].thread_ptr = XTENSA_RUR("THREADPTR");
```

Comment:
I understand this is just code motion, but frankly I think these are being stored in the wrong place.

* VECBASE is known architecturally, we should use the same static symbol address we use at CPU initialization time.

* THREADPTR is just the current thread struct, which is already stored in memory (`_kernel.cpus[_current_cpu].current`)

* EXCSAVEn registers are owned by the arch layer and are used through the abstraction layer in zsr.h, they shouldn't be touched outside arch/xtensa for any reason.  Why are we doing this?

Reply:
Coming back having looked more carefully: my guess is that we're saving EXCSAVE2/3 because we need to stash the CPU pointer register, which I don't see being restored any other way.  Yeah, don't do it this way, use the ZSR_CPU alias, that's what it's for.

Reply:
Also I was a little mixed up: threadptr is not the thread struct itself, it's the tls field of the thread struct.

And... IIRC it's not in use anywhere on SOF platforms because of the lack of support for TLS in the cadence toolchains?  Is this another spot where we're using a register we shouldn't?

Reply:
Thanks for your valuable advice. I will spend time to digest it and try to improve it.

Reply:
@andyross the HW state being saved here is also Intel implementation specific hence being in SoC specific code.